Building HPX with hwloc 2.0.0 fails

HPX fails to compile with hwloc 2.0.0 with error message: `error: use of undeclared identifier 'HWLOC_MEMBIND_REPLICATE'`. This is likely related to hwloc 2.0.0 itself and I expect it to be reproducible on any HPX setup that attempts to use it.

Compile log:
```
[1/180] Building CXX object src/CMakeFiles/hpx.dir/version.cpp.o
FAILED: src/CMakeFiles/hpx.dir/version.cpp.o 
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++  -DDEBUG -DHPX_COMPONENT_NAME=hpx -DHPX_ENABLE_ASSERT_HANDLER -DHPX_EXPORTS -D_DEBUG -D_GNU_SOURCE -Dhpx_EXPORTS -I/Users/parsa/Documents/hpx/code/hpx/repo -isystem /usr/local/include -I/usr/local/Cellar/hwloc/2.0.0/include -I. -pthread -ftemplate-depth=256 -Wall -Wextra -Wno-unused-local-typedefs -Wno-strict-aliasing -Wno-sign-promo -Wno-attributes -Wno-cast-align -Wno-unused-parameter -Wformat=2 -Wno-format-nonliteral -Winit-self -Wdouble-promotion -Wcast-qual -Wcast-align -Werror=parentheses -Werror=reorder -Werror=return-type -Werror=sequence-point -Werror=uninitialized -Werror=format -Werror=missing-braces -Werror=sign-compare -fdiagnostics-show-option -Werror=vla -Werror=return-type -g -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk -fPIC   -std=c++1z -MD -MT src/CMakeFiles/hpx.dir/version.cpp.o -MF src/CMakeFiles/hpx.dir/version.cpp.o.d -o src/CMakeFiles/hpx.dir/version.cpp.o -c /Users/parsa/Documents/hpx/code/hpx/repo/src/version.cpp
In file included from /Users/parsa/Documents/hpx/code/hpx/repo/src/version.cpp:19:
In file included from /Users/parsa/Documents/hpx/code/hpx/repo/hpx/runtime_fwd.hpp:14:
In file included from /Users/parsa/Documents/hpx/code/hpx/repo/hpx/runtime/basename_registration.hpp:14:
In file included from /Users/parsa/Documents/hpx/code/hpx/repo/hpx/lcos/future.hpp:12:
In file included from /Users/parsa/Documents/hpx/code/hpx/repo/hpx/lcos/detail/future_data.hpp:15:
In file included from /Users/parsa/Documents/hpx/code/hpx/repo/hpx/runtime/threads/thread_executor.hpp:14:
/Users/parsa/Documents/hpx/code/hpx/repo/hpx/runtime/threads/topology.hpp:70:30: error: use of undeclared identifier 'HWLOC_MEMBIND_REPLICATE'
        membind_replicate  = HWLOC_MEMBIND_REPLICATE,
                             ^
1 error generated.
[2/180] Building CXX object src/CMakeFiles/hpx.dir/state.cpp.o
FAILED: src/CMakeFiles/hpx.dir/state.cpp.o 
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++  -DDEBUG -DHPX_COMPONENT_NAME=hpx -DHPX_ENABLE_ASSERT_HANDLER -DHPX_EXPORTS -D_DEBUG -D_GNU_SOURCE -Dhpx_EXPORTS -I/Users/parsa/Documents/hpx/code/hpx/repo -isystem /usr/local/include -I/usr/local/Cellar/hwloc/2.0.0/include -I. -pthread -ftemplate-depth=256 -Wall -Wextra -Wno-unused-local-typedefs -Wno-strict-aliasing -Wno-sign-promo -Wno-attributes -Wno-cast-align -Wno-unused-parameter -Wformat=2 -Wno-format-nonliteral -Winit-self -Wdouble-promotion -Wcast-qual -Wcast-align -Werror=parentheses -Werror=reorder -Werror=return-type -Werror=sequence-point -Werror=uninitialized -Werror=format -Werror=missing-braces -Werror=sign-compare -fdiagnostics-show-option -Werror=vla -Werror=return-type -g -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk -fPIC   -std=c++1z -MD -MT src/CMakeFiles/hpx.dir/state.cpp.o -MF src/CMakeFiles/hpx.dir/state.cpp.o.d -o src/CMakeFiles/hpx.dir/state.cpp.o -c /Users/parsa/Documents/hpx/code/hpx/repo/src/state.cpp
In file included from /Users/parsa/Documents/hpx/code/hpx/repo/src/state.cpp:10:
In file included from /Users/parsa/Documents/hpx/code/hpx/repo/hpx/runtime.hpp:13:
In file included from /Users/parsa/Documents/hpx/code/hpx/repo/hpx/performance_counters/counters.hpp:11:
In file included from /Users/parsa/Documents/hpx/code/hpx/repo/hpx/lcos/future.hpp:12:
In file included from /Users/parsa/Documents/hpx/code/hpx/repo/hpx/lcos/detail/future_data.hpp:15:
In file included from /Users/parsa/Documents/hpx/code/hpx/repo/hpx/runtime/threads/thread_executor.hpp:14:
/Users/parsa/Documents/hpx/code/hpx/repo/hpx/runtime/threads/topology.hpp:70:30: error: use of undeclared identifier 'HWLOC_MEMBIND_REPLICATE'
        membind_replicate  = HWLOC_MEMBIND_REPLICATE,
                             ^
1 error generated.
[3/180] Building CXX object src/CMakeFiles/hpx.dir/exception.cpp.o
FAILED: src/CMakeFiles/hpx.dir/exception.cpp.o 
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++  -DDEBUG -DHPX_COMPONENT_NAME=hpx -DHPX_ENABLE_ASSERT_HANDLER -DHPX_EXPORTS -D_DEBUG -D_GNU_SOURCE -Dhpx_EXPORTS -I/Users/parsa/Documents/hpx/code/hpx/repo -isystem /usr/local/include -I/usr/local/Cellar/hwloc/2.0.0/include -I. -pthread -ftemplate-depth=256 -Wall -Wextra -Wno-unused-local-typedefs -Wno-strict-aliasing -Wno-sign-promo -Wno-attributes -Wno-cast-align -Wno-unused-parameter -Wformat=2 -Wno-format-nonliteral -Winit-self -Wdouble-promotion -Wcast-qual -Wcast-align -Werror=parentheses -Werror=reorder -Werror=return-type -Werror=sequence-point -Werror=uninitialized -Werror=format -Werror=missing-braces -Werror=sign-compare -fdiagnostics-show-option -Werror=vla -Werror=return-type -g -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk -fPIC   -std=c++1z -MD -MT src/CMakeFiles/hpx.dir/exception.cpp.o -MF src/CMakeFiles/hpx.dir/exception.cpp.o.d -o src/CMakeFiles/hpx.dir/exception.cpp.o -c /Users/parsa/Documents/hpx/code/hpx/repo/src/exception.cpp
In file included from /Users/parsa/Documents/hpx/code/hpx/repo/src/exception.cpp:14:
In file included from /Users/parsa/Documents/hpx/code/hpx/repo/hpx/runtime.hpp:13:
In file included from /Users/parsa/Documents/hpx/code/hpx/repo/hpx/performance_counters/counters.hpp:11:
In file included from /Users/parsa/Documents/hpx/code/hpx/repo/hpx/lcos/future.hpp:12:
In file included from /Users/parsa/Documents/hpx/code/hpx/repo/hpx/lcos/detail/future_data.hpp:15:
In file included from /Users/parsa/Documents/hpx/code/hpx/repo/hpx/runtime/threads/thread_executor.hpp:14:
/Users/parsa/Documents/hpx/code/hpx/repo/hpx/runtime/threads/topology.hpp:70:30: error: use of undeclared identifier 'HWLOC_MEMBIND_REPLICATE'
        membind_replicate  = HWLOC_MEMBIND_REPLICATE,
                             ^
1 error generated.
[4/180] Building CXX object src/CMakeFiles/hpx.dir/runtime.cpp.o
FAILED: src/CMakeFiles/hpx.dir/runtime.cpp.o 
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++  -DDEBUG -DHPX_COMPONENT_NAME=hpx -DHPX_ENABLE_ASSERT_HANDLER -DHPX_EXPORTS -D_DEBUG -D_GNU_SOURCE -Dhpx_EXPORTS -I/Users/parsa/Documents/hpx/code/hpx/repo -isystem /usr/local/include -I/usr/local/Cellar/hwloc/2.0.0/include -I. -pthread -ftemplate-depth=256 -Wall -Wextra -Wno-unused-local-typedefs -Wno-strict-aliasing -Wno-sign-promo -Wno-attributes -Wno-cast-align -Wno-unused-parameter -Wformat=2 -Wno-format-nonliteral -Winit-self -Wdouble-promotion -Wcast-qual -Wcast-align -Werror=parentheses -Werror=reorder -Werror=return-type -Werror=sequence-point -Werror=uninitialized -Werror=format -Werror=missing-braces -Werror=sign-compare -fdiagnostics-show-option -Werror=vla -Werror=return-type -g -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk -fPIC   -std=c++1z -MD -MT src/CMakeFiles/hpx.dir/runtime.cpp.o -MF src/CMakeFiles/hpx.dir/runtime.cpp.o.d -o src/CMakeFiles/hpx.dir/runtime.cpp.o -c /Users/parsa/Documents/hpx/code/hpx/repo/src/runtime.cpp
In file included from /Users/parsa/Documents/hpx/code/hpx/repo/src/runtime.cpp:11:
In file included from /Users/parsa/Documents/hpx/code/hpx/repo/hpx/performance_counters/counters.hpp:11:
In file included from /Users/parsa/Documents/hpx/code/hpx/repo/hpx/lcos/future.hpp:12:
In file included from /Users/parsa/Documents/hpx/code/hpx/repo/hpx/lcos/detail/future_data.hpp:15:
In file included from /Users/parsa/Documents/hpx/code/hpx/repo/hpx/runtime/threads/thread_executor.hpp:14:
/Users/parsa/Documents/hpx/code/hpx/repo/hpx/runtime/threads/topology.hpp:70:30: error: use of undeclared identifier 'HWLOC_MEMBIND_REPLICATE'
        membind_replicate  = HWLOC_MEMBIND_REPLICATE,
                             ^
1 error generated.
[5/180] Building CXX object src/CMakeFiles/hpx.dir/hpx_init.cpp.o
FAILED: src/CMakeFiles/hpx.dir/hpx_init.cpp.o 
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++  -DDEBUG -DHPX_COMPONENT_NAME=hpx -DHPX_ENABLE_ASSERT_HANDLER -DHPX_EXPORTS -D_DEBUG -D_GNU_SOURCE -Dhpx_EXPORTS -I/Users/parsa/Documents/hpx/code/hpx/repo -isystem /usr/local/include -I/usr/local/Cellar/hwloc/2.0.0/include -I. -pthread -ftemplate-depth=256 -Wall -Wextra -Wno-unused-local-typedefs -Wno-strict-aliasing -Wno-sign-promo -Wno-attributes -Wno-cast-align -Wno-unused-parameter -Wformat=2 -Wno-format-nonliteral -Winit-self -Wdouble-promotion -Wcast-qual -Wcast-align -Werror=parentheses -Werror=reorder -Werror=return-type -Werror=sequence-point -Werror=uninitialized -Werror=format -Werror=missing-braces -Werror=sign-compare -fdiagnostics-show-option -Werror=vla -Werror=return-type -g -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk -fPIC   -std=c++1z -MD -MT src/CMakeFiles/hpx.dir/hpx_init.cpp.o -MF src/CMakeFiles/hpx.dir/hpx_init.cpp.o.d -o src/CMakeFiles/hpx.dir/hpx_init.cpp.o -c /Users/parsa/Documents/hpx/code/hpx/repo/src/hpx_init.cpp
In file included from /Users/parsa/Documents/hpx/code/hpx/repo/src/hpx_init.cpp:11:
In file included from /Users/parsa/Documents/hpx/code/hpx/repo/hpx/apply.hpp:10:
In file included from /Users/parsa/Documents/hpx/code/hpx/repo/hpx/lcos/future.hpp:12:
In file included from /Users/parsa/Documents/hpx/code/hpx/repo/hpx/lcos/detail/future_data.hpp:15:
In file included from /Users/parsa/Documents/hpx/code/hpx/repo/hpx/runtime/threads/thread_executor.hpp:14:
/Users/parsa/Documents/hpx/code/hpx/repo/hpx/runtime/threads/topology.hpp:70:30: error: use of undeclared identifier 'HWLOC_MEMBIND_REPLICATE'
        membind_replicate  = HWLOC_MEMBIND_REPLICATE,
                             ^
1 error generated.
[6/180] Building CXX object src/CMakeFiles/hpx.dir/runtime_impl.cpp.o
FAILED: src/CMakeFiles/hpx.dir/runtime_impl.cpp.o 
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++  -DDEBUG -DHPX_COMPONENT_NAME=hpx -DHPX_ENABLE_ASSERT_HANDLER -DHPX_EXPORTS -D_DEBUG -D_GNU_SOURCE -Dhpx_EXPORTS -I/Users/parsa/Documents/hpx/code/hpx/repo -isystem /usr/local/include -I/usr/local/Cellar/hwloc/2.0.0/include -I. -pthread -ftemplate-depth=256 -Wall -Wextra -Wno-unused-local-typedefs -Wno-strict-aliasing -Wno-sign-promo -Wno-attributes -Wno-cast-align -Wno-unused-parameter -Wformat=2 -Wno-format-nonliteral -Winit-self -Wdouble-promotion -Wcast-qual -Wcast-align -Werror=parentheses -Werror=reorder -Werror=return-type -Werror=sequence-point -Werror=uninitialized -Werror=format -Werror=missing-braces -Werror=sign-compare -fdiagnostics-show-option -Werror=vla -Werror=return-type -g -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk -fPIC   -std=c++1z -MD -MT src/CMakeFiles/hpx.dir/runtime_impl.cpp.o -MF src/CMakeFiles/hpx.dir/runtime_impl.cpp.o.d -o src/CMakeFiles/hpx.dir/runtime_impl.cpp.o -c /Users/parsa/Documents/hpx/code/hpx/repo/src/runtime_impl.cpp
In file included from /Users/parsa/Documents/hpx/code/hpx/repo/src/runtime_impl.cpp:8:
In file included from /Users/parsa/Documents/hpx/code/hpx/repo/hpx/performance_counters/counters.hpp:11:
In file included from /Users/parsa/Documents/hpx/code/hpx/repo/hpx/lcos/future.hpp:12:
In file included from /Users/parsa/Documents/hpx/code/hpx/repo/hpx/lcos/detail/future_data.hpp:15:
In file included from /Users/parsa/Documents/hpx/code/hpx/repo/hpx/runtime/threads/thread_executor.hpp:14:
/Users/parsa/Documents/hpx/code/hpx/repo/hpx/runtime/threads/topology.hpp:70:30: error: use of undeclared identifier 'HWLOC_MEMBIND_REPLICATE'
        membind_replicate  = HWLOC_MEMBIND_REPLICATE,
                             ^
1 error generated.
```

## Steps to Reproduce the Problem

```sh
/usr/bin/ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"
brew install hwloc boost gperftools ninja
git clone --depth=1 https://github.com/stellar-group/hpx.git repo
mkdir build
cd build
cmake -G'Ninja' -DCMAKE_BUILD_TYPE=Debug  ../repo
cd build
ninja hello_world_exe
```

## Specifications

  - HPX Version: 922deeb0bd
  - Platform (compiler, OS): Apple Clang, macOS High Sierra
  - hwloc 2.0.0, Boost 1.66.0, gperftools 2.6.3

--- a/hpx/runtime/threads/topology.hpp
+++ b/hpx/runtime/threads/topology.hpp
@@ -69,7 +69,9 @@ namespace hpx { namespace threads
         membind_firsttouch = HWLOC_MEMBIND_FIRSTTOUCH,
         membind_bind       = HWLOC_MEMBIND_BIND,
         membind_interleave = HWLOC_MEMBIND_INTERLEAVE,
+#if HWLOC_API_VERSION < 0x00020000
         membind_replicate  = HWLOC_MEMBIND_REPLICATE,
+#endif
         membind_nexttouch  = HWLOC_MEMBIND_NEXTTOUCH,
         membind_mixed      = HWLOC_MEMBIND_MIXED,
         // special HPX addition

--- a/hpx/runtime/threads/topology.hpp
+++ b/hpx/runtime/threads/topology.hpp
@@ -71,7 +71,9 @@ namespace hpx { namespace threads
         membind_interleave = HWLOC_MEMBIND_INTERLEAVE,
         membind_replicate  = HWLOC_MEMBIND_REPLICATE,
         membind_nexttouch  = HWLOC_MEMBIND_NEXTTOUCH,
-        membind_mixed      = HWLOC_MEMBIND_MIXED
+        membind_mixed      = HWLOC_MEMBIND_MIXED,
+        // special HPX addition
+        membind_user       = HWLOC_MEMBIND_MIXED + 256
     };
 
 #include <hpx/config/warnings_prefix.hpp>

--- a/hpx/runtime/threads/topology.hpp
+++ b/hpx/runtime/threads/topology.hpp
@@ -18,6 +18,8 @@
 #include <hpx/runtime/threads/cpu_mask.hpp>
 #include <hpx/runtime/resource/partitioner_fwd.hpp>
 #include <hpx/runtime/threads/thread_data_fwd.hpp>
+#include <hpx/util/thread_specific_ptr.hpp>
+
 #include <hpx/util/spinlock.hpp>
 #include <hpx/util/static.hpp>
 
@@ -262,12 +264,12 @@ namespace hpx { namespace threads
             hpx_hwloc_membind_policy policy, int flags) const;
 
         threads::mask_type get_area_membind_nodeset(
-            const void *addr, std::size_t len, void *nodeset) const;
+            const void *addr, std::size_t len) const;
 
         bool set_area_membind_nodeset(
             const void *addr, std::size_t len, void *nodeset) const;
 
-        int get_numa_domain(const void *addr, void *nodeset) const;
+        int get_numa_domain(const void *addr) const;
 
         /// Free memory that was previously allocated by allocate
         void deallocate(void* addr, std::size_t len) const;
@@ -385,6 +387,10 @@ namespace hpx { namespace threads
         std::vector<mask_type> numa_node_affinity_masks_;
         std::vector<mask_type> core_affinity_masks_;
         std::vector<mask_type> thread_affinity_masks_;
+
+        struct tls_tag {};
+        static util::thread_specific_ptr<hpx_hwloc_bitmap_wrapper, tls_tag>
+            bitmap_storage_;
     };
 
 #include <hpx/config/warnings_suffix.hpp>

--- a/src/runtime/threads/topology.cpp
+++ b/src/runtime/threads/topology.cpp
@@ -31,6 +31,8 @@
 #include <vector>
 #include <memory>
 
+#include <errno.h>
+
 #include <hwloc.h>
 
 #if HWLOC_API_VERSION < 0x00010b00
@@ -58,58 +60,58 @@ namespace hpx { namespace threads { namespace detail
         return top.get_number_of_pus();
     }
 
-    void write_to_log(char const* valuename, std::size_t value)
-    {
-        LTM_(debug) << "topology: "
-                    << valuename << ": " << value; //-V128
-    }
-
-    void write_to_log_mask(char const* valuename, mask_cref_type value)
-    {
-        LTM_(debug) << "topology: " << valuename
-                    << ": " HPX_CPU_MASK_PREFIX
-                    << std::hex << value;
-    }
-
-    void write_to_log(char const* valuename,
-                      std::vector<std::size_t> const& values)
-    {
-        LTM_(debug) << "topology: "
-                    << valuename << "s, size: " //-V128
-                    << values.size();
-
-        std::size_t i = 0;
-        for (std::size_t value : values)
+        void write_to_log(char const* valuename, std::size_t value)
         {
-            LTM_(debug) << "topology: " << valuename //-V128
-                        << "(" << i++ << "): " << value;
+        LTM_(debug) << "topology: "
+                        << valuename << ": " << value; //-V128
         }
-    }
 
-    void write_to_log_mask(char const* valuename,
-                           std::vector<mask_type> const& values)
-    {
-        LTM_(debug) << "topology: "
-                    << valuename << "s, size: " //-V128
-                    << values.size();
-
-        std::size_t i = 0;
-        for (mask_cref_type value : values)
+        void write_to_log_mask(char const* valuename, mask_cref_type value)
         {
-            LTM_(debug) << "topology: " << valuename //-V128
-                        << "(" << i++ << "): " HPX_CPU_MASK_PREFIX
+        LTM_(debug) << "topology: " << valuename
+                        << ": " HPX_CPU_MASK_PREFIX
                         << std::hex << value;
         }
-    }
 
-    std::size_t get_index(hwloc_obj_t obj)
-    {
-        // on Windows logical_index is always -1
-        if (obj->logical_index == ~0x0u)
-            return static_cast<std::size_t>(obj->os_index);
+        void write_to_log(char const* valuename,
+            std::vector<std::size_t> const& values)
+        {
+        LTM_(debug) << "topology: "
+                        << valuename << "s, size: " //-V128
+                        << values.size();
 
-        return static_cast<std::size_t>(obj->logical_index);
-    }
+            std::size_t i = 0;
+            for (std::size_t value : values)
+            {
+            LTM_(debug) << "topology: " << valuename //-V128
+                            << "(" << i++ << "): " << value;
+            }
+        }
+
+        void write_to_log_mask(char const* valuename,
+            std::vector<mask_type> const& values)
+        {
+        LTM_(debug) << "topology: "
+                        << valuename << "s, size: " //-V128
+                        << values.size();
+
+            std::size_t i = 0;
+            for (mask_cref_type value : values)
+            {
+            LTM_(debug) << "topology: " << valuename //-V128
+                            << "(" << i++ << "): " HPX_CPU_MASK_PREFIX
+                            << std::hex << value;
+            }
+        }
+
+        std::size_t get_index(hwloc_obj_t obj)
+        {
+            // on Windows logical_index is always -1
+            if (obj->logical_index == ~0x0u)
+                return static_cast<std::size_t>(obj->os_index);
+
+            return static_cast<std::size_t>(obj->logical_index);
+        }
 }}}
 
 namespace hpx { namespace threads
@@ -1223,33 +1225,63 @@ namespace hpx { namespace threads
         hwloc_nodeset_t ns = reinterpret_cast<hwloc_nodeset_t>(nodeset);
         int ret = hwloc_set_area_membind_nodeset(topo, addr, len, ns, policy, 0);
         if (ret<0) {
+            std::string msg = std::strerror(errno);
+            if (errno == ENOSYS) msg = "the action is not supported";
+            if (errno == EXDEV)  msg = "the binding cannot be enforced";
             HPX_THROW_EXCEPTION(kernel_error
               , "hpx::threads::topology::set_area_membind_nodeset"
-              , "hwloc_set_area_membind_nodeset failed");
+              , "hwloc_set_area_membind_nodeset failed : " + msg);
             return false;
         }
         return true;
     }
 
+    util::thread_specific_ptr<hpx_hwloc_bitmap_wrapper, topology::tls_tag> topology::bitmap_storage_;
+
     threads::mask_type topology::get_area_membind_nodeset(
-        const void *addr, std::size_t len, void *nodeset) const
+        const void *addr, std::size_t len) const
     {
+        hpx_hwloc_bitmap_wrapper *nodeset = topology::bitmap_storage_.get();
+        if (nullptr == nodeset)
+        {
+            hwloc_bitmap_t nodeset_ = hwloc_bitmap_alloc();
+            topology::bitmap_storage_.reset(new hpx_hwloc_bitmap_wrapper(nodeset_));
+            nodeset = topology::bitmap_storage_.get();
+        }
+        //
         hwloc_membind_policy_t policy;
-        hwloc_nodeset_t ns = reinterpret_cast<hwloc_nodeset_t>(nodeset);
-        hwloc_get_area_membind_nodeset(topo, addr, len, ns, &policy, 0);
+        hwloc_nodeset_t ns = reinterpret_cast<hwloc_nodeset_t>(nodeset->get_bmp());
+
+        if (hwloc_get_area_membind_nodeset(topo, addr, len, ns, &policy, 0)==-1) {
+            HPX_THROW_EXCEPTION(kernel_error
+              , "hpx::threads::topology::get_area_membind_nodeset"
+              , "hwloc_get_area_membind_nodeset failed");
+            return -1;
+            std::cout << "error in  " ;
+        }
         return bitmap_to_mask(ns, HWLOC_OBJ_NUMANODE);
     }
 
-    int topology::get_numa_domain(const void *addr, void *nodeset) const
+    int topology::get_numa_domain(const void *addr) const
     {
 #if HWLOC_API_VERSION >= 0x00010b03
-        hwloc_nodeset_t ns = reinterpret_cast<hwloc_nodeset_t>(nodeset);
+        hpx_hwloc_bitmap_wrapper *nodeset = topology::bitmap_storage_.get();
+        if (nullptr == nodeset)
+        {
+            hwloc_bitmap_t nodeset_ = hwloc_bitmap_alloc();
+            topology::bitmap_storage_.reset(new hpx_hwloc_bitmap_wrapper(nodeset_));
+            nodeset = topology::bitmap_storage_.get();
+        }
+        //
+        hwloc_nodeset_t ns = reinterpret_cast<hwloc_nodeset_t>(nodeset->get_bmp());
+
         int ret = hwloc_get_area_memlocation(topo, addr, 1,  ns,
             HWLOC_MEMBIND_BYNODESET);
         if (ret<0) {
+            std::string msg(strerror(errno));
             HPX_THROW_EXCEPTION(kernel_error
-              , "hpx::threads::topology::set_area_membind_nodeset"
-              , "hwloc_set_area_membind_nodeset failed");
+              , "hpx::threads::topology::get_numa_domain"
+              , "hwloc_get_area_memlocation failed " + msg);
             return -1;
         }
         threads::mask_type mask = bitmap_to_mask(ns, HWLOC_OBJ_NUMANODE);

--- a/src/runtime/threads/topology.cpp
+++ b/src/runtime/threads/topology.cpp
@@ -1236,7 +1236,8 @@ namespace hpx { namespace threads
         return true;
     }
 
-    util::thread_specific_ptr<hpx_hwloc_bitmap_wrapper, topology::tls_tag> topology::bitmap_storage_;
+    util::thread_specific_ptr<hpx_hwloc_bitmap_wrapper, topology::tls_tag>
+        topology::bitmap_storage_;
 
     threads::mask_type topology::get_area_membind_nodeset(
         const void *addr, std::size_t len) const
@@ -1264,7 +1265,7 @@ namespace hpx { namespace threads
 
     int topology::get_numa_domain(const void *addr) const
     {
-#if HWLOC_API_VERSION >= 0x00010b03
+#if HWLOC_API_VERSION >= 0x00010b00
         hpx_hwloc_bitmap_wrapper *nodeset = topology::bitmap_storage_.get();
         if (nullptr == nodeset)
         {

